File: crs-transformation-adapter-test/src/test/java/com/programmerare/crsTransformations/compositeTransformations/CompositeStrategyTestBase.java
package com.programmerare.crsTransformations.compositeTransformations;

import com.programmerare.crsTransformationAdapterGeoPackageNGA.CrsTransformationAdapterGeoPackageNGA;
import com.programmerare.crsTransformationAdapterGeoTools.CrsTransformationAdapterGeoTools;
import com.programmerare.crsTransformationAdapterGooberCTL.CrsTransformationAdapterGooberCTL;
import com.programmerare.crsTransformationAdapterOrbisgisCTS.CrsTransformationAdapterOrbisgisCTS;
import com.programmerare.crsTransformationAdapterProj4J.CrsTransformationAdapterProj4J;
import com.programmerare.crsTransformations.CrsTransformationAdapter;
import com.programmerare.crsTransformations.coordinate.CrsCoordinate;
import com.programmerare.crsConstants.constantsByAreaNameNumber.v9_5_4.EpsgNumber;

import com.programmerare.crsTransformations.coordinate.CrsCoordinateFactory;
import org.junit.jupiter.api.BeforeAll;

import java.util.Arrays;
import java.util.List;

abstract class CompositeStrategyTestBase {

    protected static CrsTransformationAdapter adapterGeoTools;
    protected static CrsTransformationAdapter adapterGooberCTL;
    protected static CrsTransformationAdapter adapterOrbisgisCTS;
    protected static CrsTransformationAdapter adapterProj4J;
    protected static CrsTransformationAdapter adapterGeoPackageNGA;

    protected static List<CrsTransformationAdapter> allAdapters;
    protected static List<CrsCoordinate> allCoordinateResultsForTheDifferentImplementations;

    protected static double wgs84Lat = 59.330231;
    protected static double wgs84Lon = 18.059196;
    protected static double sweref99_Y_expected = 6580822;
    protected static double sweref99_X_expected = 674032;

    protected static CrsCoordinate wgs84coordinate;
    protected static CrsCoordinate resultCoordinateGeoTools;
    protected static CrsCoordinate resultCoordinateGooberCTL;
    protected static CrsCoordinate resultCoordinateOrbisgisCTS;
    protected static CrsCoordinate resultCoordinateProj4J;
    protected static CrsCoordinate resultCoordinateGeoPackageNGA;

    @BeforeAll
    final static void beforeAll() {
        adapterGeoTools = new CrsTransformationAdapterGeoTools();
        adapterGooberCTL = new CrsTransformationAdapterGooberCTL();
        adapterOrbisgisCTS = new CrsTransformationAdapterOrbisgisCTS();
        adapterProj4J = new CrsTransformationAdapterProj4J();
        adapterGeoPackageNGA = new CrsTransformationAdapterGeoPackageNGA();

        allAdapters = Arrays.asList(
            // Regarding the order of the items in the list below:
            // GeoTools should be the first since it is assumed in the test by the subclass CompositeStrategyForFirstSuccessTest
            adapterGeoTools,
            adapterGooberCTL,
            adapterOrbisgisCTS,
            adapterProj4J,
            adapterGeoPackageNGA
        );

        wgs84coordinate = CrsCoordinateFactory.createFromYNorthingLatitudeAndXEastingLongitude(wgs84Lat, wgs84Lon, EpsgNumber.WORLD__WGS_84__4326);

        resultCoordinateGeoTools = adapterGeoTools.transformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        resultCoordinateGooberCTL = adapterGooberCTL.transformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        resultCoordinateOrbisgisCTS = adapterOrbisgisCTS.transformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        resultCoordinateProj4J = adapterProj4J.transformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        resultCoordinateGeoPackageNGA = adapterGeoPackageNGA.transformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        allCoordinateResultsForTheDifferentImplementations = Arrays.asList(
            resultCoordinateGeoTools,
            resultCoordinateGooberCTL,
            resultCoordinateOrbisgisCTS,
            resultCoordinateProj4J,
            resultCoordinateGeoPackageNGA
        );
    }
}

File: Programmerare.CrsTransformations.Test/CrsTransformations/CompositeTransformations/CompositeStrategyTestBase.cs
using System.Collections.Generic;
using NUnit.Framework;
using Programmerare.CrsTransformations.Coordinate;
using Programmerare.CrsConstants.ConstantsByAreaNameNumber.v9_5_4;
using Programmerare.CrsTransformations.Adapter.DotSpatial;
using Programmerare.CrsTransformations.Adapter.ProjNet4GeoAPI;
using Programmerare.CrsTransformations.Adapter.MightyLittleGeodesy;

namespace Programmerare.CrsTransformations.CompositeTransformations 
{
public abstract class CompositeStrategyTestBase {

    protected CrsTransformationAdapterCompositeFactory crsTransformationAdapterCompositeFactory;

    protected ICrsTransformationAdapter adapterDotSpatial;
    protected ICrsTransformationAdapter adapterMightyLittleGeodesy;
    protected ICrsTransformationAdapter adapterProjNet4GeoAPI;

    protected IList<ICrsTransformationAdapter> allAdapters;
    protected List<CrsCoordinate> allCoordinateResultsForTheDifferentImplementations;

    protected const double wgs84Lat = 59.330231;
    protected const double wgs84Lon = 18.059196;
    protected const double sweref99_Y_expected = 6580822;
    protected const  double sweref99_X_expected = 674032;

    protected CrsCoordinate wgs84coordinate;
    protected CrsCoordinate resultCoordinateDotSpatial;
    protected CrsCoordinate resultCoordinateMightyLittleGeodesy;
    protected CrsCoordinate resultCoordinateProjNet4GeoAPI;

    [SetUp]
    public void SetUp() {
        crsTransformationAdapterCompositeFactory = CrsTransformationAdapterCompositeFactory.Create();

        adapterDotSpatial = new CrsTransformationAdapterDotSpatial();
        adapterProjNet4GeoAPI = new CrsTransformationAdapterProjNet4GeoAPI();
        adapterMightyLittleGeodesy = new CrsTransformationAdapterMightyLittleGeodesy();

        allAdapters = new List<ICrsTransformationAdapter>{
            // Regarding the order of the items in the list below:
            // DotSpatial should be the first since it is assumed in the test by the subclass CompositeStrategyFirstSuccessTest
            adapterDotSpatial,
            adapterProjNet4GeoAPI,
            adapterMightyLittleGeodesy
        };

        wgs84coordinate = CrsCoordinateFactory.CreateFromYNorthingLatitudeAndXEastingLongitude(wgs84Lat, wgs84Lon, EpsgNumber.WORLD__WGS_84__4326);

        resultCoordinateDotSpatial = adapterDotSpatial.TransformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        resultCoordinateProjNet4GeoAPI = adapterProjNet4GeoAPI.TransformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        resultCoordinateMightyLittleGeodesy = adapterMightyLittleGeodesy.TransformToCoordinate(wgs84coordinate, EpsgNumber.SWEDEN__SWEREF99_TM__3006);
        allCoordinateResultsForTheDifferentImplementations = new List<CrsCoordinate>{
            resultCoordinateDotSpatial,
            resultCoordinateMightyLittleGeodesy,
            resultCoordinateProjNet4GeoAPI
        };
    }
}
}